compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -814,11 +846,11 @@ object Objects:
       else
         // Widen the outer to finitize the domain. Arguments already widened in `evalArgs`.
         val (outerWidened, envWidened) =
-          if klass.owner.isClass then
+          if klass.owner.isClass then // For top-level classes, klass.owner is the enclosing package, which is a class
             (outer.widen(1), Env.NoEnv)
           else
             // klass.enclosingMethod returns its primary constructor
-            Env.resolveEnv(klass.owner.enclosingMethod, outer, summon[Env.Data]).getOrElse(Cold -> Env.NoEnv)
+            Env.resolveEnv(klass.owner.enclosingMethod, outer.asInstanceOf[ValueElement], summon[Env.Data]).getOrElse(Cold -> Env.NoEnv)
```

Comment:
Here, we know a more specific type of `outer` because we pattern matched on it.

I'd suggest renaming `value` in the pattern match case to `outer`. Then we'd have the more specific type for it.